src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,214 @@
+            /// <see cref="GetCompilationAsync"/>.
+            /// </summary>
+            [DisallowNull]
+            private Compilation? _compilationWithReplacement;
+
+            public GeneratedFileReplacingCompilationTracker(ICompilationTracker underlyingTracker, SourceGeneratedDocumentState replacementDocumentState)
+            {
+                _underlyingTracker = underlyingTracker;
+                _replacedGeneratedDocumentState = replacementDocumentState;
+            }
+
+            public bool HasCompilation => _underlyingTracker.HasCompilation;
+
+            public ProjectState ProjectState => _underlyingTracker.ProjectState;
+
+            public ICompilationTracker Clone()
+            {
+                return new GeneratedFileReplacingCompilationTracker(_underlyingTracker.Clone(), _replacedGeneratedDocumentState);
+            }
+
+            public bool ContainsAssemblyOrModuleOrDynamic(ISymbol symbol, bool primary)
+            {
+                if (_compilationWithReplacement == null)
+                {
+                    // We don't have a compilation yet, so this couldn't have came from us
+                    return false;
+                }
+                else
+                {
+                    return UnrootedSymbolSet.Create(_compilationWithReplacement).ContainsAssemblyOrModuleOrDynamic(symbol, primary);
+                }
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(Func<string, bool> predicate, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                // This only needs to be implemented if a feature that operates from a source generated file needs to search for declarations
+                // with other names; those APIs are only used by certain code fixes which isn't a need for now.
+                throw new NotImplementedException();
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(string name, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                // This only needs to be implemented if a feature that operates from a source generated file needs to search for declarations
+                // with other names; those APIs are only used by certain code fixes which isn't a need for now.
+                throw new NotImplementedException();
+            }
+
+            public ICompilationTracker Fork(ProjectState newProject, CompilationAndGeneratorDriverTranslationAction? translate = null, bool clone = false, CancellationToken cancellationToken = default)
+            {
+                // This only needs to be implemented if a feature that operates from a source generated file then makes
+                // further mutations to that project, which isn't needed for now.
+                throw new NotImplementedException();
```

Comment:
these def feel like a bunch of mines laid out for the future.  for example, if any analyzers run on sg docs and end up generatng a new doc or something...